### PR TITLE
Database

### PR DESCRIPTION
**Pull request by olgaboiar**
- Added database that stores computer's best moves calculated by minimax
- Added DB class, which is a wrapper for accessing the database
- Params for accessing the database are pulled from environmental variables for CircleCI, if no environmental variables are found, config.py reads them from database.ini
- SQL queries use SQL parameters to be SQL injection safe
- Computer's best move is pulled from DB, if not in DB is calculated by minimax
- DB has unique constrain on board state and computer marker

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -9,20 +9,42 @@ jobs:
...
+    #   - image: circleci/python:3.6.1
       
-      # Specify service dependencies here if necessary
-      # CircleCI maintains a library of pre-built images
-      # documented at https://circleci.com/docs/2.0/circleci-images/
-      # - image: circleci/postgres:9.4
+    #   # Specify service dependencies here if necessary
+    #   # CircleCI maintains a library of pre-built images
+    #   # documented at https://circleci.com/docs/2.0/circleci-images/
+    #   # - image: circleci/postgres:9.4
```
Comment (naush) [inline .circleci/config.yml]: Do you still need these comment sections?

File: .circleci/config.yml [hunk 2]
```diff
@@ -9,20 +9,42 @@ jobs:
...
+      - run: whoami
+      - run: |
+          psql \
+          -d $TEST_DATABASE_URL \
+          -c "CREATE TABLE computer_best_moves (
+            id SERIAL PRIMARY KEY,
+            board_spots VARCHAR(9) NOT NULL,
+            computer_marker VARCHAR(1) NOT NULL,
+            best_move INT);
+            ALTER TABLE computer_best_moves ADD CONSTRAINT game_state UNIQUE (board_spots, computer_marker);"
```
Comment (naush) [inline .circleci/config.yml]: Can you move this into a script file? This way the script is more reusable & we can separate changes to the script from the configuration file.

File: db.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        return params
+
+    def get_best_move_from_db(self, computer_marker, board):
+        board_spots = ''.join(map(str, board.spots))
+        query = "SELECT best_move FROM computer_best_moves WHERE computer_marker=:computer_marker AND board_spots=:board_spots"
+        move = self.postgres_db.query(query, computer_marker=computer_marker, board_spots=board_spots)
+        return move
+
+    def insert_best_move_into_db(self, computer_marker, board, best_move):
+        board_spots = ''.join(map(str, board.spots))
```
Comment (naush) [inline db.py]:
```py
board_spots = ''.join(map(str, board.spots))
```
Is duplicated on line 30 and line 24. Can you move this logic to a more appropriate class?

File: players/computer.py [hunk 4]
```diff
@@ -6,11 +6,17 @@ class Computer(Player):
...
 
     def choose_move(self, board):
         opponent = self.switch_marker(self.marker)
-        self.best_move(board, opponent)
-        return self.best_move_var
+        move = self.db.get_best_move_from_db(self.marker, board)
+
+        try:
+            return move[0].best_move
+        except IndexError:
```
Comment (naush) [inline players/computer.py]: I like to use exception with truly exceptional cases. Could we handle this without using `try...except`?

**Comment by naush** [comment]
Ideally, unit tests should not have a real connection to the database. Keep this in mind as you work on the [next story](https://trello.com/c/ixRFf08S/36-3-as-a-mentor-i-want-to-see-dependency-inversion-principle-applied-to-the-database-code-so-that-olga-has-another-opportunity-to)

File: .circleci/config.yml (hunk 1 above)
Comment (olgaboiar) [reply .circleci/config.yml]: resolved in https://github.com/olgaboiar/ttt_python/pull/8/commits/3eb32790195d129740900d935714a1938a88b035

File: db.py (hunk 3 above)
Comment (olgaboiar) [reply db.py]: resolved in https://github.com/olgaboiar/ttt_python/pull/8/commits/3eb32790195d129740900d935714a1938a88b035
- duplication removed
- a method to turn board spots array into string created in Board class

File: .circleci/config.yml (hunk 2 above)
Comment (olgaboiar) [reply .circleci/config.yml]: implemented in https://github.com/olgaboiar/ttt_python/pull/8/commits/64c76fa4568b28717bc38a465fad8c512d1914c4

File: players/computer.py (hunk 4 above)
Comment (olgaboiar) [reply players/computer.py]: refactored in https://github.com/olgaboiar/ttt_python/pull/8/commits/5cf705d6fe4c72daa53f230715ab347dc28068e9

**Pull request merged**